### PR TITLE
Expose metric for table rebalance

### PR DESCRIPTION
**Pull request by suddendust**
This PR adds a metric to track table rebalance. This metric can be useful in cases wherein we don't want certain alerts (like ingestion stopped) to trigger as a table rebalance operation is going on.

Testing steps:
- Start `RealtimeQuickStart.java` with 3 servers.
- Comment out `if(segmentAssignmentUnchanged)` block in `TableRebalancer.java` so that the dataflow actually hits the table rebalance observer.
- Rebalance the table and checkout `localhost:9021/metrics`.
- You'll see the metric (exported using the catch-all regex).

Here's the output:
 
Screenshot 2024-01-17 at 10 32 18 AM

Also, for table paused and ingestion paused, we use the catch-all regex now. Test results:

Screenshot 2024-01-17 at 5 17 51 PM

Screenshot 2024-01-17 at 5 18 05 PM

### CODE REVIEW

File: docker/images/pinot/etc/jmx_prometheus_javaagent/configs/controller.yml [hunk 1]
```diff
@@ -216,6 +216,12 @@ rules:
   labels:
     tableName: "$1"
     tableType: "$2"
+- pattern: "\"org.apache.pinot.common.metrics\"<type=\"ControllerMetrics\", name=\"pinot.controller.tableRebalanceInProgress.([^\\.]*?)_(OFFLINE|REALTIME)\"><>(\\w+)"
```
Comment (gortiz) [inline docker/images/pinot/etc/jmx_prometheus_javaagent/configs/controller.yml:219]: Do we need that? Shouldn't this be caught by the catch all rule?

File: pinot-controller/src/test/java/org/apache/pinot/controller/helix/core/rebalance/TestZkBasedTableRebalanceObserver.java [hunk 2]
```diff
@@ -41,10 +45,13 @@ void testZkObserverTracking() {
     PinotHelixResourceManager pinotHelixResourceManager = mock(PinotHelixResourceManager.class);
     // Mocking this. We will verify using numZkUpdate stat
     when(pinotHelixResourceManager.addControllerJobToZK(any(), any(), any())).thenReturn(true);
+    ControllerMetrics controllerMetrics = Mockito.mock(ControllerMetrics.class);
+    doNothing().when(controllerMetrics).setValueOfTableGauge("dummy", ControllerGauge.TABLE_REBALANCE_IN_PROGRESS, 0);
+    doNothing().when(controllerMetrics).setValueOfTableGauge("dummy", ControllerGauge.TABLE_REBALANCE_IN_PROGRESS, 1);
```
Comment (gortiz) [inline pinot-controller/src/test/java/org/apache/pinot/controller/helix/core/rebalance/TestZkBasedTableRebalanceObserver.java:50]: Do we actually need this? Isn't do nothing the default mockito behavior?

**Review by gortiz — APPROVED** [review]
Apart from my comments (which are not blockers) it looks good to me

File: docker/images/pinot/etc/jmx_prometheus_javaagent/configs/controller.yml (hunk 1 above)
Comment (suddendust) [reply docker/images/pinot/etc/jmx_prometheus_javaagent/configs/controller.yml:219]: Catch-all rule PR was just merged, so this shouldn't be needed anymore. I'll test it out.

Comment (suddendust) [reply docker/images/pinot/etc/jmx_prometheus_javaagent/configs/controller.yml]: Yes, it's working.

File: pinot-controller/src/test/java/org/apache/pinot/controller/helix/core/rebalance/TestZkBasedTableRebalanceObserver.java (hunk 2 above)
Comment (suddendust) [reply pinot-controller/src/test/java/org/apache/pinot/controller/helix/core/rebalance/TestZkBasedTableRebalanceObserver.java]: Yes, removed.

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/builder/ControllerRequestURLBuilder.java [hunk 3]
```diff
@@ -231,6 +231,10 @@ public String forTableForceCommit(String tableName) {
     return StringUtil.join("/", _baseUrl, "tables", tableName, "forceCommit");
   }
 
+  public String forTableToBrokerMappings() {
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/utils/builder/ControllerRequestURLBuilder.java:234]: Is this related?

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/ZkBasedTableRebalanceObserver.java [hunk 4]
```diff
@@ -48,23 +50,29 @@ public class ZkBasedTableRebalanceObserver implements TableRebalanceObserver {
   private boolean _isStopped = false;
   private RebalanceResult.Status _stopStatus;
 
+  private ControllerMetrics _controllerMetrics;
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/ZkBasedTableRebalanceObserver.java:53]: This can be `final`

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/ZkBasedTableRebalanceObserver.java [hunk 5]
```diff
@@ -48,23 +50,29 @@ public class ZkBasedTableRebalanceObserver implements TableRebalanceObserver {
...
+
   public ZkBasedTableRebalanceObserver(String tableNameWithType, String rebalanceJobId,
-      TableRebalanceContext tableRebalanceContext, PinotHelixResourceManager pinotHelixResourceManager) {
+      TableRebalanceContext tableRebalanceContext, PinotHelixResourceManager pinotHelixResourceManager,
+      ControllerMetrics controllerMetrics) {
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/ZkBasedTableRebalanceObserver.java:57]: No need to pass it. Use `ControllerMetrics.get()` to get the singleton instance

**Review by Jackie-Jiang — APPROVED** [review]
LGTM otherwise

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/builder/ControllerRequestURLBuilder.java (hunk 3 above)
Comment (suddendust) [reply pinot-spi/src/main/java/org/apache/pinot/spi/utils/builder/ControllerRequestURLBuilder.java:234]: Oh no, I committed thinking it was some other branch, will remove

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/ZkBasedTableRebalanceObserver.java (hunk 5 above)
Comment (suddendust) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/ZkBasedTableRebalanceObserver.java]: Done
